### ISSUE TITLE
Nullable reference types: ReferenceEquals(x, null) not considered a null check

### ISSUE THREAD

**Issue by bkoelman** [dotnet/roslyn#9930]
I have been playing with nullable reference types, which is already working remarkably great! To try it out, I have [patched up](https://github.com/bkoelman/ResharperCodeContractNullability/tree/staticnull) my analyzer to convert Resharper's `(Item)NotNull/CanBeNullAttribute` annotations to nullable reference types. Then ran that on an [actual annotated codebase](https://github.com/bkoelman/DogAgilityCompetitionManagement). 

During that process I discovered some oddities, which I'll create issues for here.

**Version Used**: Built from source, branch NullableReferenceTypes (commit https://github.com/dotnet/roslyn/commit/8553cdac75c2b0901e8ae401aaa23022f0689935)

**Steps to Reproduce**:
The next block of code raises warning 8202: Possible dereference of a null reference:

``` csharp
    public sealed class DeviceStatus
    {
        public static bool operator ==(DeviceStatus? left, DeviceStatus? right)
        {
            if (ReferenceEquals(left, right))
            {
                return true;
            }
            if (ReferenceEquals(left, null))
            {
                return false;
            }

            return left.Equals(right);   // CS8202 on 'left'
        }
    }
```

**Expected Behavior**: No CS8202 warning, because `ReferenceEquals(left, null)` ensures that `left` cannot be `null` at the `return` statement.

**Actual Behavior**:
CS8202 warning on the `left` variable in the `return` statement.

@AlekseyTs

**Comment by AlekseyTs** [comment]
@bkoelman Thanks for trying out the prototype. I agree, it makes sense to specially recognize ReferenceEquals calls.

**Comment by jcouv** [comment]
This issue was consolidated into https://github.com/dotnet/roslyn/issues/26761 (which covers a number of special methods that nullability analysis should understand, using attribute annotations).
We've implemented a few attributes, and we plan to add another one for methods like `Equals`.
I'll go ahead and close this. Thanks

**Comment by bkoelman** [comment]
While `string.IsNullOrEmpty()` is now special-cased in VS2019 Preview3, `object.ReferenceEquals()` is still not behaving as expected.

**Comment by bkoelman** [comment]
@jcouv should this be reopened or is it fixed in a recent non-public build?

**Comment by jcouv** [comment]
This issue was not fixed yet. It was consolidated into #26761 (which covers a number of special methods that nullability analysis should understand, using attribute annotations).
